rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -853,42 +855,79 @@ trait PromiseGeneration_updateAndWriteRule extends PromiseGenerationService {
   }
 
   /**
+   * For each nodeConfiguration, get the corresponding node config version.
+   * Either get it from cache or create a new one depending if the node configuration was updated
+   * or not.
+   */
+  def getNodesConfigVersion(allNodeConfigs: Map[NodeId, NodeConfiguration], hashes: Map[NodeId, NodeConfigurationHash], generationTime: DateTime): Map[NodeId, NodeConfigId] = {
+
+    /*
+     * Several steps heres:
+     * - look what node configuration are updated (based on their cache ?)
+     * - write these node configuration
+     * - update caches
+     */
+    val updatedNodes = nodeConfigurationService.selectUpdatedNodeConfiguration(allNodeConfigs, hashes)
+
+    /*
+     * The hash is directly the NodeConfigHash.hashCode, because we want it to be
+     * unique to a given generation and the "writtenDate" is part of NodeConfigurationHash.
+     * IE, even if we have the same nodeConfig than a
+     * previous one, but the expected node config was closed, we want to get a new
+     * node config id.
+     */
+    def hash(h: NodeConfigurationHash): String = {
+      //we always set date = 0, so we have the possibility to see with
+      //our eyes (and perhaps some SQL) two identicals node config diverging
+      //only by the date of generation
```

Comment:
I don't get the comment :/
So let's try to eplain the intent in an other way: today the datetime is part of the object nodeconfigurationhash. 
But for real, that object is (the hash of important things about the node) + (writtendate). We should have two subobjects, one for each group, but I didn't refactored it because it would have mean a lot more big changes. 
The written date does not reflect the configuration by itself, just the time when it is generated. And with the proposed hash format, that information is already part of the hash. 
So I chose to only took the first group about the config part in NodeConfigHash to generate a hash, and the simplest way to do so was to set the writtenDate to a dummy value (epoch is good for that).
 
Finally, all of that is completely opaque for the node and the user. We could have chosen a pur random number, it would be the same. So not sure about what mean having a difference between database and what we receive from the node. 

Does it help ?